packages/wrangler/src/deployment-bundle/node-compat.ts

```diff
@@ -0,0 +1,103 @@
+import { UserError } from "../errors";
+import { logger } from "../logger";
+
+export type NodeJSCompatMode = "legacy" | "v1" | "v2" | null;
+
+/**
+ * Validate and compute the Node.js compatibility mode we are running.
+ *
+ * Returns one of:
+ *  - "legacy": build-time polyfills, from `node_compat` flag
+ *  - "v1": nodejs_compat compatibility flag
+ *  - "v2": experimental nodejs_compat_v2 flag
+ *  - null: no Node.js compatibility
+ *
+ * Currently we require that the v2 mode is configured via `experimental:nodejs_compat_v2` compat flag,
+ * where the `experimental:` prefix is stripped before being passed to the runtime since that does not
+ * understand this prefix.
+ *
+ * We assert that only one of these modes can be specified at a time.
+ * We assert that you must prefix v2 mode with `experimental`.
+ * We warn if using legacy or v2 mode.
+ */
+export function validateNodeCompat({
+	legacyNodeCompat,
+	compatibilityFlags,
+	noBundle,
+}: {
+	legacyNodeCompat: boolean;
+	/* mutate */ compatibilityFlags: string[];
+	noBundle: boolean;
+}): NodeJSCompatMode {
+	if (legacyNodeCompat) {
+		logger.warn(
+			"Enabling Wrangler compile-time Node.js compatibility polyfill mode for builtins and globals. This is experimental and has serious tradeoffs."
+		);
+	}
+
+	const nodejsCompatV2 = compatibilityFlags.includes(
+		"experimental:nodejs_compat_v2"
+	);
+	const nodejsCompatV2NotExperimental =
+		compatibilityFlags.includes("nodejs_compat_v2");
+
+	if (nodejsCompatV2) {
+		// strip the "experimental:" prefix because workerd doesn't understand it yet.
+		compatibilityFlags[
+			compatibilityFlags.indexOf("experimental:nodejs_compat_v2")
+		] = "nodejs_compat_v2";
+	}
+
+	const nodejsCompat = compatibilityFlags.includes("nodejs_compat");
+	if (nodejsCompat && nodejsCompatV2) {
+		throw new UserError(
+			"The `nodejs_compat` and `nodejs_compat_v2` compatibility flags cannot be used in together. Please select just one."
+		);
+	}
+	if (legacyNodeCompat && nodejsCompat) {
+		throw new UserError(
+			`The ${nodejsCompat ? "`nodejs_compat`" : "`nodejs_compat_v2`"} compatibility flag cannot be used in conjunction with the legacy \`--node-compat\` flag. If you want to use the Workers ${nodejsCompat ? "`nodejs_compat`" : "`nodejs_compat_v2`"} compatibility flag, please remove the \`--node-compat\` argument from your CLI command or \`node_compat = true\` from your config file.`
+		);
+	}
+
+	if (legacyNodeCompat && (nodejsCompat || nodejsCompatV2)) {
+		throw new UserError(
+			`The ${nodejsCompat ? "`nodejs_compat`" : "`nodejs_compat_v2`"} compatibility flag cannot be used in conjunction with the legacy \`--node-compat\` flag. If you want to use the Workers ${nodejsCompat ? "`nodejs_compat`" : "`nodejs_compat_v2`"} compatibility flag, please remove the \`--node-compat\` argument from your CLI command or \`node_compat = true\` from your config file.`
+		);
```

Comment:
This seems duplicative?

Reply:
Agh! Cut and paste mistake. Good spot.

Reply:
https://github.com/cloudflare/workers-sdk/pull/5991/commits/852db02ad3c8ff499c5910180c037b5a137f2f1d